File: test/GSS.Authentication.CAS.E2E.Tests/PageObjects/LoginPage.cs
namespace GSS.Authentication.CAS.E2E.Tests.PageObjects;

public class LoginPage
{
    private readonly IPage _page;

    public LoginPage(IPage page)
    {
        _page = page;
    }

    public ILocator Heading => _page.GetByRole(AriaRole.Heading, new() { Level = 1 });
    public ILocator CasButton => _page.Locator("button, a").GetByText("CAS", new() { Exact = true });
    public ILocator OpenIdConnectButton => _page.Locator("button, a").GetByText("OpenIdConnect", new() { Exact = true });

    public async Task<string> GetHeadingTextAsync()
    {
        return await Heading.TextContentAsync() ?? string.Empty;
    }

    public async Task SelectCasAsync()
    {
        await CasButton.ClickAsync();
    }

    public async Task SelectOpenIdConnectAsync()
    {
        await OpenIdConnectButton.ClickAsync();
    }

    public async Task<bool> HasAuthenticationSchemesAsync(int timeout = 10000)
    {
        try
        {
            await CasButton.Or(OpenIdConnectButton).WaitForAsync(new() { State = WaitForSelectorState.Visible, Timeout = timeout });
        }
        catch (System.Exception)
        {
            // Ignore timeout
        }
        var casVisible = await CasButton.IsVisibleAsync();
        var oidcVisible = await OpenIdConnectButton.IsVisibleAsync();
        return casVisible || oidcVisible;
    }
}

File: e2e/support/login-page.ts
import type { Locator, Page } from '@playwright/test';

export class LoginPage {
  readonly heading: Locator;
  readonly casButton: Locator;
  readonly openIdConnectButton: Locator;

  constructor(private readonly page: Page) {
    this.heading = page.getByRole('heading', { level: 1 });
    this.casButton = page.locator('button, a').getByText('CAS', { exact: true });
    this.openIdConnectButton = page.locator('button, a').getByText('OpenIdConnect', { exact: true });
  }

  async selectCas(): Promise<void> {
    await this.casButton.click();
  }

  async selectOpenIdConnect(): Promise<void> {
    await this.openIdConnectButton.click();
  }

  async hasAuthenticationSchemes(timeout = 10000): Promise<boolean> {
    await this.casButton
      .or(this.openIdConnectButton)
      .waitFor({ state: 'visible', timeout })
      .catch(() => undefined);
    return (await this.casButton.isVisible()) || (await this.openIdConnectButton.isVisible());
  }
}
